content/languages/cpp/authoring/index.md

```diff
@@ -0,0 +1,387 @@
+size_t input_length = gen_length(); // use length generator
+for(size_t i=0; i<input_length; ++i) {  
+  // string generation logic...
+  input.push_back(gen_letter()); // use generator of random letters
+}
+
+// ...testing logic    
+```
+
+The standard library also contains some functions which can be helpful when generating random inputs:
+
+- [`std::shuffle`](https://www.cplusplus.com/reference/algorithm/shuffle/) - randomly rearranges elements in range using generator.
+- [`std::sample`](https://en.cppreference.com/w/cpp/algorithm/sample) - randomly selects elements from the sequence, without replacement.
+
+Note that `std::random_shuffle` is now considered obsolete and has been superseded by `std::shuffle`.
+
+### Reference solution
+
+If the test suite happens to use a reference solution to calculate expected values (which [should be avoided](/authoring/guidelines/submission-tests/#reference-solution) when possible), or some kind of reference data like precalculated arrays, etc., it must not be possible for the user to call it, redefine, overwrite, or directly access its contents. To prevent this, reference data can be made a private member of the `Describe` structure. A reference solution can be either a private member of `Describe`, or a lambda-initialized local variable of an `It` block. See `[Example test suite](#example-test-suite)` for some examples.
+
+The reference solution or data ___must not___ be defined in the [Preloaded code](/authoring/guidelines/preloaded/).
+
+
+### Input mutation
+
+General guidelines for submission tests contain a section related to [input mutation](/authoring/guidelines/submission-tests/#input-mutation) and how to prevent users from abusing it to work around kata requirements. It's usually not a problem in "real world" C++ programming, but on Codewars, users can take advantage of vulnerable test suites and modify their behavior. Constness of a function argument can be forcefully cast away by a user, or they can change the function signature and remove the `const` qualifier from input parameter(s).  After calling a user solution, tests should not rely on the state of such values, and they should consider them as potentially modified by the user.
+
+
+### Calling assertions
+
+Snowhouse provides a set of useful [assertions](/languages/cpp/igloo#snowhouse-assertions)`, but when used incorrectly, they can cause a series of problems:
+- Use of an assertion not suitable for the given case may lead to incorrect test results,
+- Incorrectly used assertions may produce confusing or unhelpful messages.
+
+To avoid the above problems, calls to assertion functions should respect the following rules:
+- The expected value should be calculated _before_ invoking an assertion. The `expected` parameter passed to the assertion should not be a function call expression, but a value calculated directly beforehand. _(? does this point hold for C++ kata?)_
```

Comment:
> does this point hold for C++ kata

If it's about hiding the name of the reference solution, then it depends on whether it can be called from users' solutions. If it's about input mutation, it depends on whether the input is copied for the reference solution. If everything is done the best way, it shouldn't matter.

Reply:
```suggestion
```

I think will remove it for now, and we will see if reviewers will be able to handle issues related to both above aspects correctly. Or maybe just make this point a recommendation instead of `should`?